pac4cli/wpad.py

```diff
@@ -129,27 +130,57 @@ def get_config_wpad_url(self, config_file):
             logger.info("No wpad url specified")
             return None
 
+    @inlineCallbacks
+    def get_dns_wpad_urls(self, domains):
+        dns_urls = []
+        for domain in domains:
+            # We need to be aware of the TLD in the passed domains, so as to 
+            # avoid looking for the wpad url outside the business/company/entity
+            # e.g.: for http://hostname.subdomain.example.co.uk, we want to try: 
+            # http://wpad.hostname.subdomain.example.co.uk/wpad.dat
+            # http://wpad.subdomain.example.co.uk/wpad.dat
+            # http://wpad.example.co.uk/wpad.dat
+            # and avoid trying:
+            # http://wpad.co.uk/wpad.dat
+            #
+            # The tld package is using Mozilla's database for top level domains.
+            logger.info("Found fqdn: '%s'", domain)
+            tld_res = yield get_tld(domain, as_object=True, fix_protocol=True)
+            logger.debug("tld_res.subdomain: %s", tld_res.subdomain)
+            logger.debug("tld_res.fld: %s", tld_res.fld)
+            domain_parts = tld_res.subdomain.split('.')
+            for i in range(1, len(domain_parts) + 1):
+                level_domain = '.'.join(domain_parts[i:])
+                wpad_search_domain = tld_res.fld
+                if level_domain:
```

Comment:
If I understand correctly, we can avoid this `if` condition by having `i` range only until `len(domain_parts)` (exclusive) instead of `len(domain_parts)+1`. If I'm not mistaken, `level_domain` is the empty string iff `domain_parts[i:]` is an empty list, which happens if `i` is at least `len(domain_parts)`.

Reply:
consider the fqdn `my-pc.example.com`. And we did `range(1, len(domain_parts)`:

the `domain_parts` will be of len = 1, which will skip the loop all together, which results in empty list, while we should expect one url: `wpad.example.com/wpad.dat`
But we can make it work by also moving out the shortest url from being populated at the last loop to be populated directly after exiting the loop. after all we already know it will be a candidate (`wpad.example.com/wpad.dat`).

Reply:
Ah I see. I hadn't noticed that `wpad_search_domain` is being assigned to in two different places, and only one of them is guarded by the `if`.

Can I suggest this alternative? Seems to be a bit more clear to me:
```python
subdomain_parts = tld_res.subdomain.split('.')
# we're taking end slices of zero or more subdomain parts
for i in range(len(subdomain_parts)+1):
     parts = ["wpad"] + subdomain_parts[len(subdomain_parts)-i:] + [tld_res.fld]
     wpad_search_domain = ".".join(parts)
     dns_urls.append("http://{}/wpad.dat".format(wpad_search_domain))
```
Does this have the same behaviour as what you're after or am I missing something?

Reply:
This is what I ended up doing:
```
for i in range(1, len(domain_parts)):
    level_domain = '.'.join(domain_parts[i:])
    wpad_search_domain = '.'.join([level_domain, tld_res.fld])
    logger.debug("appending: '%s'", "http://wpad.{}/wpad.dat".format(wpad_search_domain))
    dns_urls.append("http://wpad.{}/wpad.dat".format(wpad_search_domain))
dns_urls.append("http://wpad.{}/wpad.dat".format(tld_res.fld))
```
But I have to say, yours is more lean! .. at least joining all in one `join`. I still need to get used to the python way of dealing with lists

Reply:
There's also a functional difference: my code gives `len(subdomain_parts) + 1` items in `dns_urls`, whereas yours gives `len(domain_parts)` items (and that includes the one you append after the loop). Correct?

The difference seems to be the one obtained by adding `wpad.` in front of the _full_ domain, e.g.

    http://wpad.hostname.subdomain.example.co.uk/wpad.dat

in the example from the code comment.

You could get that behaviour in your version by replacing `range(1, len(domain_parts))` by `range(len(domain_parts))`.

Reply:
You are right, this is indeed a difference. But still, the behaviour is correct, and the my comment is wrong. the search for `wpad.hostname.subdomain.example. com/wpad.dat` should not be included.

I'm sorry I caused this confusion with my slopyness in the comment.

Reply:
Nevertheless, I'm adopting your version (only fixing the start index). It is more elegant :)